### ISSUE TITLE
gst.Structure needs copying after transfer:none

### ISSUE THREAD

**Issue by danijelk00** [go-gst/go-gst#129]
After fixing the issues  #115 and #122 , I encountered new problems while extracting information from the rtpsession element.

My code looks like this:

```
for{
	stats, err := rtpSession.GetProperty("stats")
 	if err != nil {
	  logger.Fatal("%v Error getting stats:", direction, err)
	}
	if statsStructure, ok = stats.(*gst.Structure); ok {
		stats_string := statsStructure.String()
	}
}

```
I want to extract certain values from the rtpSession statistics, so I decided to convert the gst.Structure into a string in order to further extract the necessary information using regex.

However, I occasionally get warnings such as (probably when calling String()):

<img width="932" alt="gst-assertions" src="https://github.com/user-attachments/assets/6af87ff9-56a9-4b9c-b849-210cec81097e">

and after a certain amount of time, the program crashes.

**Comment by RSWilli** [comment]
From looking at it, this looks like an early GC of the `GValue`, before you are done with the structure. Could you please experiment with calling `runtime.GC()` between these lines. I hope that this will remove the "occasionally" and "after a certain amount of times" from the reproduction.

The goal should be that this crashes every time.

**Comment by danijelk00** [comment]
Yes, now I am getting warnings in every iteration. It seems that the return value from **rtpSession.GetProperty("stats")** is freed prematurely, which is causing my issues.

**Comment by RSWilli** [comment]
Okay that confirms my suspicion, but the fixing these kinds of issues is annoying. I will check it when I have time.

In the meantime you could look into it yourself, or disable automatic GC in your application.

**Comment by RSWilli** [comment]
I checked the rust bindings for reference:

* The `FromValue` used `get_boxed` https://gitlab.freedesktop.org/gstreamer/gstreamer-rs/-/blob/main/gstreamer/src/structure.rs#L327-329
* `from_glib_none` uses `gst_structure_copy` https://gitlab.freedesktop.org/gstreamer/gstreamer-rs/-/blob/main/gstreamer/src/structure.rs#L270

I think the fix for your issue is simply copying the structure before returning in the marshal here: https://github.com/go-gst/go-gst/blob/main/gst/gst_wrappers.go#L464-L468

@danijelk00 if you want you can create a PR for this.

**Comment by danijelk00** [comment]
Thank you so much for your time.
I will test if it works on Monday and will give you feedback then.

**Comment by RSWilli** [comment]
After thinking a bit more about this, this is also an issue with messages, I reported in #31 but got it wrong there, structures are not refcounted, but they are not owned by you when they get passed via `transfer:none`. There might be other cases where a copy is needed.

Also if we copy the structure, then a finalizer that frees the copy will also be necessary.

**Comment by RSWilli** [comment]
@danijelk00 check the linked PR please

**Comment by danijelk00** [comment]
@RSWilli  
That looks much better! 

When I test locally on a simple program, although I call runtime.GC() in several places, the program runs normally and I don't get any warnings. 
On Monday, I will test it on the much larger application I am working on and let you know if everything is fine, so that we can close this issue.

**Comment by danijelk00** [comment]
@RSWilli 
I just tested it, and it works very well! Can you commit the latest changes?

**Comment by RSWilli** [comment]
> This warning popped up only once, but the program doesn't crash and works normally.
**GStreamer-CRITICAL : 09:20:49.747: gst_structure_free: assertion 'GST_STRUCTURE_REFCOUNT (structure) == NULL' failed

That's still a memory corruption :/ I'll check this out later this week.

**Comment by RSWilli** [comment]
this should be fixed in current main, merged through #153
